File: apps/server/src/auth/users/service.rs
use crate::auth::*;
use crate::shared::AppResult;

use std::sync::Arc;
use sword::prelude::*;

#[injectable]
pub struct UsersService {
	users: Arc<UsersRepository>,
	hasher: Arc<Hasher>,
}

impl UsersService {
	pub async fn find(&self, query: GetUsersQuery) -> AppResult<Vec<UserView>> {
		let filter = UserFilter {
			role: query.role,
			search: query.search,
		};

		self.users.list(filter).await
	}

	pub async fn find_by_id(&self, id: &UserId) -> AppResult<UserView> {
		let Some(user) = self.users.find_by_id(id).await? else {
			return Err(AuthError::UserNotFound)?;
		};

		Ok(UserView::from(user))
	}

	pub async fn create(&self, mut dto: CreateUserDto) -> AppResult<UserView> {
		if self.users.find_by_email(&dto.email).await?.is_some() {
			Err(AuthError::EmailAlreadyExists)?;
		}

		dto.password = self.hasher.hash(&dto.password)?;

		let user = self.users.create(&dto).await?;

		Ok(UserView::from(user))
	}

	pub async fn update(&self, id: &UserId, mut dto: UpdateUserDto) -> AppResult<UserView> {
		let Some(user) = self.users.find_by_id(id).await? else {
			return Err(AuthError::UserNotFound)?;
		};

		if let Some(ref email) = dto.email
			&& email != &user.email
			&& self.users.find_by_email(email).await?.is_some()
		{
			Err(AuthError::EmailAlreadyExists)?;
		}

		if let Some(ref password) = dto.password {
			dto.password = Some(self.hasher.hash(password)?);
		}

		let user = self.users.update(&user.id, &dto).await?;

		Ok(UserView::from(user))
	}

	pub async fn delete(&self, id: &UserId) -> AppResult<()> {
		if self.users.find_by_id(id).await?.is_none() {
			Err(AuthError::UserNotFound)?;
		}

		self.users.delete(id).await
	}
}

File: apps/client/src/lib/users/service.ts
import { User } from "./entity"
import { http } from "$lib/shared/http/client"

import type { CreateUserDTO, UpdateUserDTO, UserDTO } from "./dtos"

class UsersService {
	public list(params?: { search?: string; role?: string }): Promise<User[]> {
		const users = http.request<UserDTO[]>({
			method: "GET",
			url: "/users",
			params,
		})

		return users.then((users) => users.map(User.fromDTO))
	}

	public get(id: string): Promise<User> {
		return http.request<User>({
			method: "GET",
			url: `/users/${id}`,
		})
	}

	public create(data: CreateUserDTO): Promise<User> {
		return http.request<User>({
			method: "POST",
			url: "/users",
			data,
		})
	}

	public update(id: string, data: UpdateUserDTO): Promise<User> {
		return http.request<User>({
			method: "PUT",
			url: `/users/${id}`,
			data,
		})
	}

	public delete(id: string): Promise<void> {
		return http.request<void>({
			method: "DELETE",
			url: `/users/${id}`,
		})
	}
}

export const usersService = new UsersService()
